[지하철노선도 / Step1] 로이(김덕우) 미션 제출합니다

the9kim opened the pull request:
안녕하세요.

페어프로그래밍으로 구현한 핵심 도메인 로직을 간단하게 설명드리고자 합니다.

1. 'Section' 객체를 이용하여 지하철 역 구간(현재역 - 도착역)  구간을 표현했습니다.

2. 지하철 노선은 연결리스트와 유사한 구조로 구현했습니다.
    - 지하철 노선(Line) 객체는 연결리스트 Head에 해당하는 구간(Section)객체를 필드값으로 가집니다.
    - 개별 구간(Section) 객체는 다음에 연결되는 구간 객체(NextSection)를 필드값으로 가집니다.

스프링을 처음 배우는 입장이라 부족한 점이 많습니다.

개선할 점을 알려주시면 충분히 고민하고 반영해보겠습니다.

감사합니다!

java-saeng committed:
remove : 기존 스켈레톤 코드 삭제

java-saeng committed:
feat : 이미 존재하는 Section의 중간이나 끝에 새로운 Section을 추가하는 기능 구현

java-saeng committed:
feat : Line에 새로운 Section을 넣을 때, starter 가 바뀌는 경우 기능 구현

java-saeng committed:
feat : 중간 Section 삭제 기능 구현

java-saeng committed:
feat : starter 삭제 기능 구현

java-saeng committed:
feat : starter를 delete 하는 기능 구현

java-saeng committed:
feat : 스키마 설정

java-saeng committed:
feat : Line 에 존재하는 역들을 조회하는 기능 구현

java-saeng committed:
feat : 역 삭제/추가 기능 구현

java-saeng committed:
feat : Line의 마지막 역 삭제 기능 구현

java-saeng committed:
refactor : Line 메서드 extract 및 test 수정

java-saeng committed:
refactor : Section 메서드 extract 및 테스트 수정

java-saeng committed:
refactor : 도메인에 id 인스턴스 필드 삭제

java-saeng committed:
feat : Line 에 Section 이 한개만 있을 경우 역을 삭제하면 Line 도 지워져야한다.

java-saeng committed:
feat : 새로운 노선 생성 기능 구현

java-saeng committed:
refactor : Section Dto 로 변환하는 map 메서드를 SectionService 로 이동

java-saeng committed:
fix : Line이 삭제되는 경우 조건 추가

java-saeng committed:
feat : Distance 값 객체 추가

begaonnuri commented on src/main/java/subway/controller/LineController.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    @GetMapping("/lines")
+    public List<LineResponse> searchAllSectionInLines(
```
메서드 명과 반환 타입이 다른 부분이 있네요. 의도한대로 반환 타입을 수정하거나 메서드명을 수정해보면 어떨까요?

the9kim replied:
노선을 조회하는 메서드 이름에 Section이 포함되니 의미전달이 제대로 안되는 듯 합니다. 

해당 메서드명을 'SearchLines'로 변경해봤습니다!

begaonnuri commented on src/main/java/subway/controller/LineController.java:
```diff
@@ -0,0 +1,36 @@
+
+    @GetMapping("/lines")
+    public List<LineResponse> searchAllSectionInLines(
+            @RequestBody(required = false) SearchAllSectionLineRequest searchAllSectionLineRequest
```
Get 요청인데 RequestBody를 사용하신 이유가 있을까요? 다음 링크도 참고해보세요 🙂

https://developer.mozilla.org/ko/docs/Web/HTTP/Methods/GET

the9kim replied:
하나의 컨트롤러 메서드를 이용하여 '전체 노선 조회'와 '노선 하나 조회' 기능을 동시에 구현하려다 보니 RequestBody를 이용하게 됐습니다.

그런데 굳이 하나의 메서드로 두가지 기능을 동시에 수행할 필요는 없을 듯 하여 Api를 두개(1.전체 노선 조회, 2. 하나의 노선 조회)로 다시 구분했습니다.

또 Get 요청 시, 노선의 Id값(@PathVariable)을 통해 해당 노선을 조회하는 방식으로 수정해봤습니다!

begaonnuri commented on src/main/java/subway/controller/LineController.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    @PostMapping("/lines")
+    public void registerLine(
```
생성 요청 후 반환이 없으면 클라이언트에서 추가적인 작업이 어렵지 않을까요?

예를들어 `게시글 생성 이후 해당 게시글의 상세 페이지를 보여준다` 와 같은 요구사항이 있을 수도 있을 것 같아요. 이를 위해 응답 Body나 Location 헤더를 통해 자원의 위치를 표현해주면 어떨까요?

the9kim replied:
생성에 대한 응답 추가가 필요하다는 말에 동의합니다!

노선을 조회하는 페이지로 리다이렉트 하도록 반환값을 수정해봤습니다!

begaonnuri replied:
음.. 제가 말씀드린 예시는 단순 예시이고 백엔드보단 프론트엔드의 요구사항일 수 있어요.

다만 프론트엔드든, API 사용자든 생성 이후 조회에 사용할 수 있도록 생성한 자원을 Body로 응답해주거나 Location 헤더를 통해 자원의 위치를 응답하는 REST API를 구성해보면 좋겠다는 의도였습니다!

따라서 리다이렉트보단 Body 혹은 Location 헤더를 통해 표현해보면 어떨까요?

https://restfulapi.net/http-status-201-created/

the9kim replied:
제가 잘못 이해했군요..ㅎㅎ

말씀하신대로 Location 헤더로 자원의 위치를 표시해봤습니다.

ResponseEntity에 관해서 같이 공부해보고 적용해봤습니다!

begaonnuri commented on src/main/java/subway/controller/StationController.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    @DeleteMapping("/stations")
+    public void deleteStation(@RequestBody StationDeleteRequest stationDeleteRequest) {
```
`DELETE` 요청의 경우 RequestBody 대신 삭제할 리소스를 path variable를 통해 식별합니다.

각각 어떤 경우에 path variable, request param, request body를 사용할까요? 질문에 답을 적어주시면서 학습해보세요!

the9kim replied:
1. @RequestParam: Get 쿼리 파라미터 전송이나 Post HTML Form 전송 시 요청 파라미터를 조회하기 위해 사용합니다.
2. @PathVariable: parameter 형식이 아닌 경로로 전달되는 변수를 조회할 경우 사용합니다.
3. @RequestBody: HTTP 메시지 바디 정보를 조회할 경우 사용합니다.

아직까지 Api를 구성하는 방법이 익숙치가 않아서 연습이 많이 필요할 듯 합니다ㅠ

말씀하신대로 PathVariable을 사용하여 Delete요청을 처리하는 방식으로 수정해봤습니다!

begaonnuri commented on src/main/java/subway/dao/LineDao.java:
```diff
@@ -2,60 +2,57 @@
+    private final JdbcTemplate jdbcTemplate;
+    private final SimpleJdbcInsert simpleJdbcInsert;
+    private final RowMapper<LineEntity> rowMapper = (rs, rowNum) ->
+            new LineEntity(
```
Line과 LineEntity를 분리하신 이유가 궁금합니다!

the9kim replied:
지하철 노선 구성을 연결리스트와 유사한 자료구조로 만드는 과정에서,

 Line 객체는 Head값으로 첫번 째 구간(Section) 정보를 가지고,

개별 구간 객체는 다음에 이어지는 구간 객체의 참조값을 가지도록 구현했습니다.

DB를 설계할 때 Line의 헤더나, Section의 다음 섹션값을 테이블 필드로 넣어서 구현하는 과정이 복잡해서 해당 내용은 생략 하게 되었습니다. 

결국 DB에 저장한 데이터로 Line 객체를 바로 만드는 것이 어려워 LineEntity 객체를 별도로 분리했습니다!

begaonnuri commented on src/main/java/subway/dao/LineDao.java:
```diff
@@ -2,60 +2,57 @@
-                .usingGeneratedKeyColumns("id");
+        this.simpleJdbcInsert =
+                new SimpleJdbcInsert(jdbcTemplate)
+                        .withTableName("LINE")
```
테이블 이름만 컨벤션이 다른 이유가 있을까요?

the9kim replied:
특별한 이유는 없었습니다ㅠ

컨벤션을 통일하는 것이 좋을 듯 해서,

테이블명과 필드명을 모두 소문자로 변경해봤습니다!

begaonnuri commented on src/main/java/subway/dao/SectionEntity.java:
```diff
@@ -0,0 +1,49 @@
+    private int distance;
+    private Long lineId;
+
+    public SectionEntity(final Long id,
```
불필요한 내용이 길어지는데 this()를 통해 생성자 체이닝을 사용하면 어떨까요?

the9kim replied:
생성자 체이닝이라는 방식을 처음 알았습니다ㅠ

생성자가 여러개일 경우 훨씬 깔끔한 코드를 구현할 수 있군요!

말씀하신 방식대로 수정해봤습니다!

begaonnuri commented on src/main/java/subway/domain/Distance.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    public Distance minus(final Distance other) {
+        return new Distance(value - other.value);
```
값 객체를 불변으로 잘 유지해주셨네요 👍🏻 equals와 hashCode 재정의도 좋습니다!

begaonnuri commented on src/test/java/subway/domain/LineTest.java:
```diff
@@ -0,0 +1,266 @@
+
+class LineTest {
+
+    /**
+     * 기존 : A -> B
+     * 추가 : B -> C
+     * 결과 : A -> B -> C
+     */
+    @Test
```
주석으로 표현해주셔서 이해하기 훨씬 쉽네요 👍🏻

begaonnuri commented on src/main/java/subway/domain/Section.java:
```diff
@@ -0,0 +1,107 @@
+package subway.domain;
+
+public class Section {
```
Section 객체를 적절하게 도출해주신것같네요!

begaonnuri commented on src/main/java/subway/domain/Station.java:
```diff
@@ -1,41 +1,22 @@
 
-import java.util.Objects;
-
 public class Station {
```
Station도 Distance처럼 불변으로 유지하면 어떨까요?

불변으로 지정 여부를 결정하는 로이만의 기준이 있었나요?

the9kim replied:
노선 구성을 연결리스트로 구현하다 보니,

역을 삽입하거나 삭제하는 과정에서 기존 섹션에 포함된 역의 이름을 변경해야 하는 경우가 있었습니다.

setter를 쓰지 않기 위해 Station을 새로 생성하는 방법을 고민했지만, 

station의 일급 컬렉션인 section이 다른 section과 참조 관계로 엮여 있는 탓에 쉽지가 않았습니다ㅠ

그래서 결국 Station에서 Setter(updateStationName())를 활용하는 방법을 결정하게 됐습니다!

begaonnuri commented on src/main/java/subway/service/dto/RegisterLineRequest.java:
```diff
@@ -0,0 +1,25 @@
+
+public class RegisterLineRequest {
+
+    private String currentStationName;
```
final 사용이 혼재되어있는데 자바는 언어차원에서 불변을 지원하지 않기 때문에 명시적으로 표현해주는것이 중요합니다!

the9kim replied:
네 값 변경이 필요없는 경우 불변 객체를 사용하는 습관을 들여야겠네요!

Dto와 Entity 중 불변이 아닌 객체를 불변으로 변경해봤습니다.

begaonnuri commented on src/main/java/subway/service/CommonService.java:
```diff
@@ -0,0 +1,51 @@
+
+import java.util.List;
+
+@Component
+public class CommonService {
```
CommonService의 역할은 무엇인가요? Service 대신 Component를 사용하신 이유도 궁금합니다!

the9kim replied:
CommonService는 역 삽입과 삭제 로직을 수행할 때, DB에 저장된 데이터를 기반으로 Line 객체를 만드는 역할을 수행합니다.

CommonService라는 명칭이 좀 애매한 듯 하여, 'LineMakerService'로 변경해봤습니다. 

@Component를 사용한건 단순 실수며, @Service로 변경했습니다ㅜ

begaonnuri replied:
다른 서비스와 이 서비스의 역할이 미묘하게 다른것 같지 않나요?

다른 서비스는 DB와 같은 외부 서비스를 통해 비즈니스 로직을 이어가지만 이 서비스는 `객체 생성`에 초점이 맞춰져 있는 것 같네요. 이런 경우 서비스가 아닌 또 다른 계층임을 암시하기도 합니다😊 아직 학습을 추천드리진 않으니 이 부분은 천천히 고민해보세요!

begaonnuri reviewed (changes requested):
어려운 도메인이었을텐데 잘 구현해주셨네요.
몇가지 리뷰 남겼으니 확인해주세요.

the9kim committed:
refactor: 노선 조회 기능 메서드 이름 수정

the9kim committed:
refactor: 노선 생성 기능 반환값 수정

- 기존: 반환 값 미존재
- 변경: 노선 생성 후 조회 페이지로 이동하도록 반환값 변경

the9kim committed:
refactor: sql 테이블 및 필드명 컨벤션 통일

- 기존: 대소문자 혼용
- 변경: 소문자로 통일

the9kim committed:
refactor: 생성자 체이닝을 활용하여 Entity 내 중복 코드 제거

the9kim committed:
refactor: Dto 및 Entity 불변 객체로 변경

the9kim committed:
refactor: 어노테이션 변경(Component -> service) 및 클래스명 변경

the9kim committed:
refactor: 노선 조회 API 분리 및 리소스 변경

기존: 전체 노선 조회 및 단일 노선 조회를 하나의 컨트롤러 메서드로 구현 + RequestBody 이용

변경: 전체 노선 조회와 단일 노선 조회 API 구분 + 단일 노선 조회 시 노선 Id값(@PathVariable) 이용

the9kim committed:
refactor: 역 삭제 관련 API 수정

기존: Delete 메서드 맵핑 시, @RequestBody 사용

변경: @PathVariable을 이용하여 삭제 리소스 식별

the9kim commented:
안녕하세요 터틀!

학습이 필요한 점 위주로 많이 알려주셔서 감사합니다!

사실 마음이 급해서 차분히 고민하지 못한 감이 있지만..

그래도 코멘트 주신 의도를 생각해보고 수정이 필요하다고 판단한 부분을 고쳐봤습니다!

바쁘신 와중에 소중한 코멘트 남겨주셔서 감사합니다👍🏻

begaonnuri commented on src/main/java/subway/controller/StationController.java:
```diff
@@ -17,13 +15,13 @@ public StationController(final StationService stationService) {
-    @DeleteMapping("/stations")
-    public void deleteStation(@RequestBody StationDeleteRequest stationDeleteRequest) {
-        stationService.deleteStation(stationDeleteRequest);
+    @DeleteMapping("/{id}/{station}")
```
이런 경우 station의 id로 착각할 여지가 있지 않을까요? lineId가 적절할것같아요.

또한 의미가 잘 드러나지 않는것같은데 `특정 라인에 있는 특정 역`이라는 의미를 담아 변경해보면 어떨까요?

the9kim replied:
동의합니다!

라인에 포함 역이라는 의미를 나타내도록 네이밍을 변경해봤습니다!

begaonnuri commented on src/main/java/subway/service/dto/RegisterLineRequest.java:
```diff
@@ -1,11 +1,25 @@
+    private final String lineName;
+    private final int distance;
+
+    @JsonCreator
+    public RegisterLineRequest(@JsonProperty(value = "stationName") String currentStationName,
+                               @JsonProperty(value = "nextStationName") String nextStationName,
+                               @JsonProperty(value = "lineName") String lineName,
+                               @JsonProperty(value = "distance") int distance) {
```
JsonCreator 또는 JsonProperty 둘 중 하나만 사용해도 식별이 가능합니다.

the9kim replied:
알려주셔서 감사합니다!

JsonCreator는 삭제했습니다!

begaonnuri reviewed (approved):
리뷰 잘 반영해주셨네요 로이!
몇가지 코멘트 남겼는데 2단계에서 함께 반영해주세요.
그리고 도메인에 대한 단위 테스트만 진행해주셨는데요, 웹 미션인 만큼 로이가 적절하다고 생각하는 테스트 방법을 통해 웹 테스트도 함께 작성해주세요.
이번 단계는 여기서 머지하겠습니다. 고생하셨습니다 👍🏻
